TST: add array-agnostic assertions

lucascolley opened the pull request:
#### Reference issue
This comment from @mdhaber: https://github.com/scipy/scipy/pull/18930#discussion_r1304458805

#### What does this implement/fix?
Three new assertions in `_lib._array_api.py`:
```python
assert_equal(actual, desired, err_msg='', xp=None)

assert_close(actual, desired, rtol=None, atol=None, err_msg='', xp=None)`

assert_less(actual, desired, err_msg='', verbose=True, xp=None)`
```

Here, the appropriate testing function is used from `numpy`, `cupy` or `torch` based on:
1. If given, the `xp` parameter.
2. Otherwise, `array_namespace(actual)`.

#### Additional information
These functions are just to enable GPU testing internally in SciPy, hopefully concisely.

`torch` does not have an `assert_array_less` equivalent as far as I can see, so I have transferred to cpu and used `numpy.testing`.

lucascolley commented on scipy/_lib/_array_api.py:
```diff
@@ -163,3 +163,53 @@ def copy(x, *, xp=None):
+        return xp.testing.assert_array_equal(actual, desired, err_msg=err_msg)
+    elif 'torch' in xp.__name__:
+        return xp.testing.assert_close(actual, desired, rtol=0, atol=0,
+                                       msg=err_msg)
```
from https://pytorch.org/docs/stable/testing.html:
![image](https://github.com/scipy/scipy/assets/51488791/3a19abcd-3d4c-4297-9d81-7277e427b4a5)

rgommers replied:
Funky but recommended indeed. You may want to add a comment in the line above to avoid the next reader of this code being puzzled. Something like:
```
# PyTorch recommends using `rtol=0, atol=0` like this to test for exact equality
```

lucascolley commented:
cc @tylerjereddy @rgommers since this approach is a compromise between the two methods we discussed with `assert_allclose` on #19005: using device transfers (once we decide how to) where necessary, but avoiding them where possible by making use of `cupy.testing` and `torch.testing`.

tylerjereddy commented on scipy/_lib/_array_api.py:
```diff
@@ -163,3 +163,53 @@ def copy(x, *, xp=None):
+
+
+def assert_less(actual, desired, err_msg='', verbose=True, xp=None):
+    # TODO compare arrays from different libraries
```
How does this gel with `_assert_matching_namespace` from: https://github.com/scipy/scipy/pull/19005/files#diff-b440e60b73eceb9148588107815b1b5cd66d6858e3f56f8d4872584b704a02e7R172 ?

Or maybe there's a concrete example in mind where we aren't ready to strictly require the namespace preservation?

lucascolley replied:
The motivation for this is to be able to, say, write `assert_equal(x, y)` where `x` is an `xp` array and `y` is an `ndarray`. This purely makes writing the tests easier. Our public functions should still avoid device transfers - this is just a convenience to make the process of modifying tests easier.

As for how it gels with `_assert_matching_namespace`, I don't see any issues. Couldn't we still use it immediately after the function we are testing? Or, as I have it in #19005, just in a separate class.

rgommers replied:
I think I'd be happy to requite matching namespaces, unless there's a good reason why this won't suffice:

> Perhaps it is better to always call `xp.asarray` from within the test though

lucascolley replied:
Also noting that, if we do require the namespaces to match, `array_namespace(actual, desired)` should be used.

lucascolley replied:
> Is the idea that instead of converting from xp to np to perform the comparison with np.testing, we convert from np to xp and perform the comparison with the xp's testing function?

Yes.

> If that's right, I take it that xp.asarray is guaranteed to accept a np array as input (for any Array API compatible xp), even if np.asarray is not guaranteed to work on an xp array? (Otherwise, we'd just reversed the original problem.)

I think so. [`xp.asarray`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html) accepts any object which supports the Python buffer protocol. There isn't a clear answer from google but [this page](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html) seems to suggest that `numpy.ndarray`s do support it.

> I think that separating the details of array backends from the intent of the test would be nice. As the test author, I would just want these functions to compare the data in the arrays (and possibly shape and dtype - something that I think is missing in some of the np.testing functions), regardless of backend.

Totally agree with this. Abstracting as much as possible out to these assertions seems desirable.

> If xp.asarray is guarranteed to work with np input, requiring the conversion in the test would suffice, but if it's going to be done in every test, it would be more convenient to just put it inside the function.

I agree. If Tyler and Ralf are happy with this then we can start to think about which library's assertions to use for each combination of arrays, and how to implement this.

lucascolley replied:
> I think so. xp.asarray accepts any object which supports the Python buffer protocol. 

Just learned that the buffer protocol is CPU only, so it seems not.

mdhaber replied:
> Just learned that the buffer protocol is CPU only, so it seems not.

It sounds like this reinforces the need for these assertions to do the conversions themselves, then.

lucascolley replied:
https://data-apis.org/array-api/2022.12/design_topics/data_interchange.html#data-interchange It looks like we should be using DLPack. Maybe we can build this into our `as_xparray` helper.

lucascolley replied:
I checked to see what scikit-learn are doing: https://github.com/scikit-learn/scikit-learn/blob/a05eb6bbe30bc40904274c87afb4ac93bcd60168/sklearn/model_selection/tests/test_split.py#L1286-L1335

They are also just using `xp.asarray` to go from `np` to `xp`. It's true that this is not guaranteed to work for any library, but it's sufficient while we only test with libraries for which it works. I don't know how likely it is that we will want to test with a library for which this doesn't work.

For `assert_allclose`, they are converting back to `np` using `_convert_to_numpy`, which has the necessary shims for when `np.asarray` doesn't work: https://github.com/scikit-learn/scikit-learn/blob/2a2772a87b6c772dc3b8292bcffb990ce27515a8/sklearn/utils/_array_api.py#L191-L204.

rgommers replied:
tl;dr `xp.asarray(a_numpy_array)` should be robust enough, and about the best we can do.

More detailed: the buffer protocol is CPU-only indeed. So CuPy does something special to make it efficient, converting directly from NumPy. But worst-case, if a library does not support NumPy arrays directly )_and_ also not the buffer protocol, then a numpy array should still get treated like a Python list (or list of lists), and iterated over - extracting the data element by element. The `asarray` spec contains `NestedSequence` as an input, so that is supposed to work. There is no more generic option.

I can't think of a current real-world scenario where this wouldn't work. If a library doesn't support CPUs at all, it'd have much bigger problems of course, so I don't think we have to care about such hypothetical corner cases.

rgommers replied:
> (and possibly shape and dtype - something that I think is missing in some of the `np.testing` functions)

+1. I just noticed this earlier when looking at Lucas's `fft` PR: PyTorch is more strict here, so a direct translation of tests from numpy to array API was failing here.

tylerjereddy commented:
> These functions are just to enable GPU testing internally in SciPy, hopefully concisely.

In case it helps, I started writing tests for these tests (yeah, I know...) a bit locally to help review.

Small patch to this branch with new tests below the fold (not fully portable yet):

<details>

```diff
diff --git a/scipy/_lib/tests/meson.build b/scipy/_lib/tests/meson.build
index 197bb3f28..56042ed87 100644
--- a/scipy/_lib/tests/meson.build
+++ b/scipy/_lib/tests/meson.build
@@ -13,6 +13,7 @@ python_sources = [
   'test_tmpdirs.py',
   'test_warnings.py',
   'test_array_api.py',
+  'test_gpu_compare.py',
 ]
 
 py3.install_sources(
diff --git a/scipy/_lib/tests/test_gpu_compare.py b/scipy/_lib/tests/test_gpu_compare.py
new file mode 100644
index 000000000..4e8484323
--- /dev/null
+++ b/scipy/_lib/tests/test_gpu_compare.py
@@ -0,0 +1,44 @@
+from scipy._lib._array_api import assert_equal, assert_close, assert_less
+
+import numpy as np
+import pytest
+try:
+    import cupy as cp
+    import torch
+    HAS_BACKENDS = 1
+except ImportError:
+    HAS_BACKENDS = 0
+
+
+@pytest.mark.skipif(not HAS_BACKENDS,
+                    reason="GPU backends not available")
+@pytest.mark.parametrize("arr", [
+    torch.tensor([[-7.07, 3.5], [1, 2.9090]],
+                 device=torch.device('cuda:0')),
+    cp.array([[11.7, 3.6], [0.01, 99]]),
+    np.ones((4, 4)),
+    ])
+def test_true_matches(arr):
+    # check that identical arrays are compared
+    # as we'd expected using our built-in
+    # array API shims on GPU
+    assert_equal(arr, arr)
+    assert_close(arr, arr)
+    with pytest.raises(AssertionError):
+        assert_less(arr, arr)
+
+
+@pytest.mark.skipif(not HAS_BACKENDS,
+                    reason="GPU backends not available")
+@pytest.mark.parametrize("arr", [
+    torch.tensor([[-7.07, 8.999], [-5.0, 9.0]],
+                 device=torch.device('cuda:0')),
+    cp.array([[11.7, 3.6], [11.8, 3.601]]),
+    np.array([[11.7, 3.6], [11.8, 3.601]]),
+    ])
+def test_less_case(arr):
+    assert_less(arr[0], arr[1])
+    with pytest.raises(AssertionError):
+        assert_equal(arr[0], arr[1])
+    with pytest.raises(AssertionError):
+        assert_close(arr[0], arr[1])
```

</details>

The tests then pass with `SCIPY_ARRAY_API=1 python dev.py test -j 32 -t scipy/_lib/tests/test_gpu_compare.py`, which I suppose is a good sign. I don't know that it makes sense to add those tests or not, but perhaps something like that if we're going to use them in a few places, though if they're GPU-focused probably can only run them periodically.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -163,3 +163,53 @@ def copy(x, *, xp=None):
+                                          atol=atol, err_msg=err_msg)
+    elif 'torch' in xp.__name__:
+        return xp.testing.assert_close(actual, desired, rtol=rtol,
+                                       atol=atol, msg=err_msg)
```
The `atol` defaults for `torch.testing.assert_close` are not great. How about using an explicit `atol=0` for everything in this unified testing function?

lucascolley replied:
The defaults for every library are now `rtol=1e07, atol=0`.

rgommers replied:
LGTM, resolving

rgommers commented:
Looks good and useful to me, thanks Lucas. And thanks for the tests Tyler, +1 for adding those.

tupui reviewed (approved):
I think we can proceed. This seems helpful already and as it will only be used internally by us, we can address other use cases later.

I propose to merge this week.

lucascolley commented:
Okay, if we are going to merge without the device-transfers in these assertions, I'll remove the `TODO` comments.

> I think that separating the details of array backends from the intent of the test would be nice. As the test author, I would just want these functions to compare the data in the arrays (and possibly shape and dtype - something that I think is missing in some of the np.testing functions), regardless of backend.
> 
> If xp.asarray is guarranteed to work with np input, requiring the conversion in the test would suffice, but if it's going to be done in every test, it would be more convenient to just put it inside the function.

This should be moved to a follow-up issue if not addressed here. I haven't found the `xp.asarray`s in test functions to be particularly laborious, but it would probably be an improvement to have them inside the assertions instead. We'd have to decide on which assertion to use for each combination of array libraries.

rgommers commented on scipy/_lib/_array_api.py:
```diff
@@ -163,3 +163,45 @@ def copy(x, *, xp=None):
+
+
+def assert_close(actual, desired, rtol=1e-07, atol=0, err_msg='', xp=None):
+    # TODO compare arrays from different libraries
```
can you remove this comment? I think we're going the way of comparing arrays from the same library here, right?

rgommers replied:
doh, you wrote that an hour ago.

rgommers commented:
Agreed with @tupui that this seems good to get merged, it's ready as is it seems (modulo comment tweaks). 

> > If xp.asarray is guarranteed to work with np input, requiring the conversion in the test would suffice, but if it's going to be done in every test, it would be more convenient to just put it inside the function.
> 
> This should be moved to a follow-up issue if not addressed here. I haven't found the `xp.asarray`s in test functions to be particularly laborious, but it would probably be an improvement to have them inside the assertions instead. We'd have to decide on which assertion to use for each combination of array libraries.

We shouldn't have to guess about what assertion to use. If it's ambiguous, I'd prefer explicit over implicit. I don't think it is though? We should never convert the `actual` input, only the `desired` one.

In many cases we may construct `desired` directly from function/constructor calls though, so it's not completely clear to me that an extra `xp.asarray` call inside test functions is a win. But I think it'll be apparent quickly enough once someone actually tries this.

mdhaber commented on scipy/_lib/_array_api.py:
```diff
@@ -163,3 +163,45 @@ def copy(x, *, xp=None):
+    # TODO compare arrays from different libraries
+    if xp is None:
+        xp = array_namespace(actual)
+    if 'cupy' in xp.__name__:
```
There was some discussion about how to do this in https://github.com/scipy/scipy/pull/19023#discussion_r1286902145 and https://github.com/scipy/scipy/pull/19023#discussion_r1295885319. Since it looks like this will merge before gh-19005, do you want to move the `is_numpy` and other helpers into this PR?

lucascolley replied:
Yep can do. We probably want to match both the 'raw' and `array_api_compat` versions in that case, since lots of tests don't stick to `(actual, desired)` and have `(desired, actual)` instead, where the desired is raw but the actual will be the compat version. E.g https://github.com/scipy/scipy/blob/6da65af8dac2cf9ea4b94a4dc8e73dd7ffe27225/scipy/fft/tests/test_fftlog.py#L95-L108

lucascolley commented:
I think that I've addressed all comments now, apart from Tyler's tests. Shall I add those in as well (with co-author)?

CI seems to be having issues, but looks unrelated?

tupui commented:
I am fine with follow ups as this is ready and a blocker for other WIP.

mdhaber commented:
I'll go ahead and merge. We can adjust as needed.
